harness/determined/exec/tensorboard.py

```diff
@@ -1,143 +1,280 @@
-    """Return True if the process successfully comes up before a deadline."""
+def get_tensorboard_args(tb_version, tfevents_dir, add_args):
+    """Build tensorboard startup args.
 
-    deadline = time.time() + max_seconds
+    Args are added and deprecated at the mercy of tensorboard; all of the below are necessary to
+    support versions 1.14, 2.4, and 2.5
 
-    while True:
-        if time.time() > deadline:
-            print(f"TensorBoard did not find metrics within {max_seconds} seconds", file=sys.stderr)
-            return False
+    - Tensorboard 2+ no longer exposes all ports. Must pass in "--bind_all" to expose localhost
+    - Tensorboard 2.5.0 introduces an experimental feature (default load_fast=true)
+    which prevents multiple plugins from loading correctly.
+    """
+    task_id = os.environ["DET_TASK_ID"]
+    port = os.environ["TENSORBOARD_PORT"]
 
-        if not still_alive_fn():
-            print("TensorBoard process died before reporting metrics", file=sys.stderr)
-            return False
+    tensorboard_args = [
+        "tensorboard",
+        f"--port={port}",
+        f"--path_prefix=/proxy/{task_id}",
+        *add_args
+    ]
 
-        time.sleep(1)
+    # Version dependant args
+    version_parts = tb_version.split(".")
+    major = int(version_parts[0])
+    minor = int(version_parts[1])
 
-        try:
-            res = requests.get(url)
-            res.raise_for_status()
-        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError):
-            continue
+    if major >= 2:
+        tensorboard_args.append("--bind_all")
+    if major > 2 or major == 2 and minor >= 5:
+        tensorboard_args.append("--load_fast=false")
 
-        try:
-            tags = res.json()
-        except ValueError:
-            continue
+    tensorboard_args.append(f"--logdir={tfevents_dir}")
+
+    return tensorboard_args
+
+
+def get_tensorboard_url():
+    task_id = os.environ["DET_TASK_ID"]
+    port = os.environ["TENSORBOARD_PORT"]
+    tensorboard_addr = f"http://localhost:{port}/proxy/{task_id}"
+    return f"{tensorboard_addr}/data/plugin/scalars/tags"
+
+
+class S3Connector:
+    def __init__(self, storage_config):
+        req_keys = ["bucket", "access_key", "secret_key", "endpoint_url"]
+        for key in req_keys:
+            try:
+                storage_config[key]
+            except KeyError:
+                raise ValueError(f"storage_config must define a '{key}'")
+
+        self.client = boto3.client(
+            "s3",
+            endpoint_url=storage_config["endpoint_url"],
+            aws_access_key_id=storage_config["access_key"],
+            aws_secret_access_key=storage_config["secret_key"],
+        )
+
+        self.delimiter = "/"
+        self._s3_path_regex = re.compile(r"^s3://(?P<bucket>[^/]+)/(?P<key>.+)$")
+        self._path_type = re.compile(r"^(?P<type>[^:]+):/(?P<path>/.+)$")
+
+    def _path_to_bucket_key(self, s3_path):
```

Comment:
I struggled with how to deal with the paths once inside this code. My thoughts were:
1. Keep everything as `s3://<bucket>/<key>` type paths (for consistency)
2. Just pass the `<key>`s around inside of this code, since, as you pointed out, we do already know about the bucket.
I am currently just using this function to validate that my path matches what I expect, and to pull the key part out.

The likely will go away when we change what we pass in to something like:
`{ "path_to_load": [ {"storage_config": { ... }, "keys": [...]}, ... }`